eppo_core/src/event_ingestion/delivery.rs

```diff
@@ -50,108 +46,88 @@ impl DeliveryStatus {
+    // We use this unbounded channel to loop back messages that need retrying.
+    let (retries_tx, mut retries_rx) =
+        mpsc::unbounded_channel::<(/* attempts: */ u32, BatchedMessage<Event>)>();
+
     loop {
-        let BatchedMessage {
-            batch,
-            flush: _flush,
-        } = uplink.recv().await?;
-        if batch.is_empty() {
-            continue;
+        // Randomly select between sending a new batch or retrying an old one.
+        let (attempts, msg) = tokio::select! {
+            msg = uplink.recv() => (0, msg?),
+            msg = retries_rx.recv() => msg?,
+        };
+
+        let BatchedMessage { batch, flush } = msg;
+
+        let mut result = event_delivery.deliver(batch).await;
+
+        if attempts >= config.max_retries {
+            // Exceeded max retries -> promote retriable errors to permanent ones.
+            result.failure.append(&mut result.retry);
         }
-        let events_to_deliver = batch
-            .iter()
-            .map(|queued_event| &queued_event.event)
-            .collect::<Vec<&Event>>();
-        let result = event_delivery.deliver(events_to_deliver.as_slice()).await;
-        match result {
-            Ok(response) => {
-                let DeliveryStatus {
-                    success,
-                    failure,
-                    retry,
-                } = collect_delivery_response(batch, response, max_retries);
-                // forward successful and failed events to the delivery status channel
-                if !success.is_empty() || !failure.is_empty() {
-                    delivery_status
-                        .send(DeliveryStatus::new(success, failure, Vec::new()))
-                        .await
-                        .ok()?;
-                }
-                // forward retryable events to the retry channel
-                if !retry.is_empty() {
-                    retry_downlink
-                        .send(BatchedMessage::new(retry, None))
-                        .await
-                        .ok()?;
-                }
-            }
-            Err(err) => {
-                match err {
-                    EventDeliveryError::RetriableError(_) => {
-                        // Retry later
-                        retry_downlink
-                            .send(BatchedMessage::new(batch, None))
-                            .await
-                            .ok()?;
-                    }
-                    _ => {
-                        warn!("Failed to deliver events: {}", err);
-                        // In this case there is no point in retrying delivery since the error is
-                        // non-retriable.
-                        delivery_status
-                            .send(DeliveryStatus::failure(batch))
-                            .await
-                            .ok()?;
-                    }
-                }
-            }
+
+        let retry_batch = std::mem::take(&mut result.retry);
```

Comment:
`std::mem::take` replaces `result.retry` with an empty vector and returns all events to us (indiana jones style).

This way, we avoid cloning the whole batch of events for a retry but as a side effect, this means that retried events are not included in the delivery status (which nobody reads anyway for now)